frontend/modules/tree/treeProcessing/processing.ts

```diff
@@ -0,0 +1,431 @@
+import { memoize, intersection, keys, pickBy, uniqBy } from 'lodash';
+import { SELECTION_STATES, VISIBILITY_STATES, NODE_TYPES } from '../../../constants/tree';
+
+export class Processing {
+	public nodesList = [];
+	public nodesIndexesMap = {};
+	public defaultVisibilityMap = {};
+	public meshesByModelId = {};
+	public nodesBySharedIdsMap = {};
+	public selectionMap = {};
+	public visibilityMap = {};
+
+	constructor(data) {
+		const {
+			nodesList, nodesIndexesMap, defaultVisibilityMap,
+			meshesByModelId, nodesBySharedIdsMap, visibilityMap, selectionMap
+		} = data;
+		this.nodesList = nodesList;
+		this.nodesIndexesMap = nodesIndexesMap;
+		this.defaultVisibilityMap = defaultVisibilityMap;
+		this.meshesByModelId = meshesByModelId;
+		this.visibilityMap = visibilityMap;
+		this.selectionMap = selectionMap;
+		this.nodesBySharedIdsMap = nodesBySharedIdsMap;
+	}
+
+	public get selectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
```

Comment:
> I'm guessing the difference between them is one returns only selected nodes, and one returns parents of selected as well as selected?
Yes, exactly.

> maybe highlightedNodesIds should be called fullySelectedIds?
The idea was to set that name because it represents the current state of viewer - nodes returned by that getter are highlighted on the viewer. :) `selected` means they are selected in some way e.g on the nodes list.
But yes, We can change the name. No problem ;)

Reply:
✅